src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,159 @@
+            head = nodeToRemove.next;
+        } else if (nodeToRemove == tail) {
+            nodeToRemove.previous.next = null;
+            tail = nodeToRemove.previous;
+        } else {
+            nodeToRemove.previous.next = nodeToRemove.next;
+            nodeToRemove.next.previous = nodeToRemove.previous;
+        }
+        nodeToRemove.next = nodeToRemove.previous = null;
+        nodeToRemove.storedItem = null;
+        --size;
+    }
+
+    private Node<T> findNode(int index) {
+        Node<T> headHandler = head;
+        Node<T> tailHandler = tail;
+        if (index < size / 2) {
+            for (int i = 0; i < index; i++) {
+                headHandler = headHandler.next;
+            }
+            return headHandler;
+        } else {
+            for (int i = size - 1; i > index; i--) {
+                tailHandler = tailHandler.previous;
+            }
+            return tailHandler;
+        }
+    }
+
     @Override
     public boolean add(T value) {
-        return false;
+        Node<T> nodeToAdd = new Node<>(tail, value, null);
+        if (head == null) {
+            head = tail = nodeToAdd;
+        } else {
+            tail.next = nodeToAdd;
+            tail = nodeToAdd;
+        }
+        size++;
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index == size) {
+            add(value);
+            return;
+        }
+        checkBounds(index);
+        insertNode(findNode(index), value);
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (T element : list) {
+            add(element);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkBounds(index);
+        return findNode(index).storedItem;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkBounds(index);
+        Node<T> nodeInPlace = findNode(index);
+        T elementToReplace = nodeInPlace.storedItem;
+        nodeInPlace.storedItem = value;
+        return elementToReplace;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkBounds(index);
+        Node<T> nodeToRemove = findNode(index);
+        T previousValue = nodeToRemove.storedItem;
+        unlink(nodeToRemove);
+        return previousValue;
     }
 
     @Override
     public boolean remove(T t) {
+        Node<T> holder = head;
+        for (int i = 0; i < size; i++) {
+            if (holder.storedItem == t || holder.storedItem.equals(t)) {
+                unlink(holder);
+                return true;
+            }
+            holder = holder.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private static class Node<T1> {
```

Comment:
That's optional, but why T1 and not just T ?)

Reply:
done